Allow YAML list format for config options

jdavcs opened the pull request:
Fixes #10942

This allows specifying multiple values in galaxy.yml as a YAML list in addition to comma-separated string:

```yaml
tool_data_table_config_path:
- path1
- path2

some_other_config: 'path1, path2'
```
EDIT: This is applicable to any options that allow multiple values (not just paths). This includes `admin_users`.

jdavcs committed:
Allow YAML list format for config path options

jdavcs committed:
Enable list format for admin_users option

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -274,13 +274,17 @@ def convert_datatype(key, value):
             if value is not None and datatype in type_converters:
-                return type_converters[datatype](value)
+                # convert value or each item in value to type `datatype`
+                if type(value) is list:
```
```suggestion
                if isinstance(value, list):
```

nsoranzo commented on lib/galaxy/config/__init__.py:
```diff
@@ -274,13 +274,17 @@ def convert_datatype(key, value):
-                return type_converters[datatype](value)
+                # convert value or each item in value to type `datatype`
+                if type(value) is list:
+                    return [type_converters[datatype](item) for item in value]
```
Assigning `type_converters[datatype]` to a variable (before the `if`) would make this more readable, I think.

jdavcs replied:
Agreed. Done.

jdavcs committed:
Use isinstance() instead of type()

Co-authored-by: Nicola Soranzo <nicola.soranzo@gmail.com>

jdavcs committed:
Improve readability as per code review

nsoranzo commented on test/unit/config/test_path_resolves_to.py:
```diff
@@ -209,6 +209,15 @@ def test_kwargs_listify(mock_init, monkeypatch):
+    new_path4 = ['new1', 'new2']
+    config = BaseAppConfiguration(path4=new_path4)
+
+    assert config._raw_config['path4'] == 'new1,new2'
```
This surprised me a bit, I expected this to still be a list, but it is joined into a comma-separated string at https://github.com/ic4f/galaxy/blob/dev_yaml_list_format/lib/galaxy/config/__init__.py#L298 , but only for paths (`admin_users` doesn't have `resolves_to`).

Not sure if this can become a problem at some point, but I thought I should mention it.

jdavcs replied:
I saw it too; and now that you mention this, I think we shouldn't leave it as is: the issue here is that now the `strip_deprecated_dir` function has this unexpected side effect (for list values).  Which, apart from being bad design, can later lead to problems, of course.
I'll fix it. I won't be ideal, but it'll preserve the `value` data type. 
In general, I think the config module is becoming too complicated. I think, if we have a way to mark all path values in the schema (making `path-resolves_to` mandatory for paths is one way), the logic can be simplified considerably. I should get to it soon.

jdavcs replied:
@nsoranzo  Could you take a look at the latest commit?

jdavcs committed:
Do not change list value to csv string

Remove side effect from strip_deprecated_dir function.
For context, see code review.
